language/move-lang/src/unit_test/plan_builder.rs

```diff
@@ -0,0 +1,297 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    cfgir::ast as C,
+    expansion::ast as E,
+    shared::{known_attributes, Address, CompilationEnv, Identifier},
+    unit_test::{Context, ExpectedFailure, ModuleTestPlan, TestCase},
+};
+use move_core_types::{account_address::AccountAddress as MoveAddress, value::MoveValue};
+use move_ir_types::location::Loc;
+use std::collections::BTreeMap;
+
+//***************************************************************************
+// Test Plan Building
+//***************************************************************************
+
+// Constructs a test plan for each module in `prog`. This also validates the structure of the
+// attributes as the test plan is constructed.
+pub fn construct_test_plan(
+    compilation_env: &mut CompilationEnv,
+    prog: &C::Program,
+) -> Option<Vec<ModuleTestPlan>> {
+    if !compilation_env.flags().testing_set() {
+        return None;
+    }
+    let mut context = Context::new(compilation_env);
+    Some(
+        prog.modules
+            .iter()
+            .flat_map(|(_, module_ident, module_def)| {
+                construct_module_test_plan(&mut context, module_ident, module_def)
+            })
+            .collect(),
+    )
+}
+
+fn construct_module_test_plan(
+    context: &mut Context,
+    module_ident: &(Address, String),
+    module: &C::ModuleDefinition,
+) -> Vec<ModuleTestPlan> {
+    let tests: BTreeMap<_, _> = module
+        .functions
+        .iter()
+        .filter_map(|(loc, fn_name, func)| {
+            build_test_info(context, loc, fn_name, func)
+                .map(|test_case| (fn_name.to_string(), test_case))
+        })
+        .collect();
+
+    if tests.is_empty() {
+        vec![]
+    } else {
+        vec![ModuleTestPlan::new(module_ident, tests)]
+    }
+}
+
+fn build_test_info(
```

Comment:
All of this is probably fine for now, but it feels very ad-hoc in that anything else that uses attributes in the future is likely going to have to duplicate a lot of this. 

We should maybe have some schema system for specifying properties over known attributes

Reply:
I think this would be awesome to have, but we may want to wait on it and I'll do it in a followup PR if that sounds good?

Reply:
I'm not sure it is needed anytime soon. Definitely something we can wait on